tests/src/test/java/alluxio/client/cli/fs/ForceMasterSyncAfterReadOutOfSyncTest.java

```diff
@@ -0,0 +1,388 @@
+import alluxio.multi.process.PortCoordination;
+import alluxio.server.ft.journal.raft.EmbeddedJournalIntegrationTestBase;
+import alluxio.util.CommonUtils;
+import alluxio.util.WaitForOptions;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.ratis.server.RaftServerConfigKeys;
+import org.apache.ratis.server.storage.RaftStorageImpl;
+import org.apache.ratis.statemachine.impl.SimpleStateMachineStorage;
+import org.apache.ratis.statemachine.impl.SingleFileSnapshotInfo;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+public class ForceMasterSyncAfterReadOutOfSyncTest
+    extends EmbeddedJournalIntegrationTestBase {
+
+  private static final int RESTART_TIMEOUT_MS = 6 * Constants.MINUTE_MS;
+  private static final int NUM_MASTERS = 3;
+  private static final int NUM_WORKERS = 0;
+  private final InstancedConfiguration mConf = Configuration.copyGlobal();
+
+  public class TestFileInStream extends AlluxioFileInStream {
+
+    protected TestFileInStream(URIStatus status,
+                               InStreamOptions options,
+                               FileSystemContext context) {
+      super(status, options, context);
+    }
+  }
+
+  @Test
+  public void forceMasterSync() throws Exception {
+    mCluster = MultiProcessCluster.newBuilder(PortCoordination.EMBEDDED_JOURNAL_FAILOVER)
+        .setClusterName("ForceMasterSyncAfterReadOutOfSyncTest_forceMasterSync")
+        .setNumMasters(NUM_MASTERS)
+        .setNumWorkers(NUM_WORKERS)
+        .addProperty(PropertyKey.MASTER_JOURNAL_TYPE, JournalType.EMBEDDED)
+        .addProperty(PropertyKey.MASTER_JOURNAL_FLUSH_TIMEOUT_MS, "5min")
+        .addProperty(PropertyKey.MASTER_EMBEDDED_JOURNAL_MIN_ELECTION_TIMEOUT, "750ms")
+        .addProperty(PropertyKey.MASTER_EMBEDDED_JOURNAL_MAX_ELECTION_TIMEOUT, "1500ms")
+        .build();
+    mCluster.start();
+
+    FileSystem fs = mCluster.getFileSystemClient();
+    AlluxioURI testFile = new AlluxioURI("/test.txt");
+    FileOutStream outStream = fs.createFile(testFile);
+    String fileContent = "testing";
+    outStream.write(fileContent.getBytes());
+    outStream.close();
+
+    URIStatus uriStatus = fs.getStatus(testFile);
+
+    OpenFilePOptions readOptions =
+        OpenFilePOptions.newBuilder().setReadType(ReadPType.CACHE_PROMOTE).build();
+
+    TestFileInStream testFileInStream = new TestFileInStream(uriStatus,
+        new InStreamOptions(uriStatus, readOptions, mConf),
+        mCluster.getFilesystemContext());
+    try {
+      testFileInStream.read();
+    } catch (Exception e) {
+      assertTrue(e instanceof IllegalStateException);
+      testFileInStream.read();
+    }
```

Comment:
```suggestion
// The file has been changed in the UFS and Alluxio does not know
// So reading on the stream will fail because the length doesn't meet expectation
Exception e = Assert.assertThrows(IllegalStateException.class, () -> {
  testFileInStream.read();
}
assertTrue(e.getMessage().contains("blah"));
// The failed stream will notify the master to sync on the next read
// The next read will just succeed
testFileInStream.read();
```